fl4health/losses/mkmmd_loss.py

```diff
@@ -77,156 +80,218 @@ def construct_quadruples(self, X: torch.Tensor, Y: torch.Tensor) -> torch.Tensor
-        # For the quadruples of the form (x,  x', y, y') we need distances for (x, x'), (y, y'), (x, y'), (x, y')
-        # NOTE: The inner product for a vector x^Tx is just the square and sum of the components of x
-        x_x_prime = (
-            torch.sum((v_i_quadruples[:, 0, :] ** 2), dim=1).reshape(-1, 1)
-            + torch.sum((v_i_quadruples[:, 1, :] ** 2), dim=1).reshape(1, -1)
-            - 2.0 * torch.mm(v_i_quadruples[:, 0, :], torch.transpose(v_i_quadruples[:, 1, :], 0, 1))
-        )
-        x_x_prime[x_x_prime < 0] = 0
-        y_y_prime = (
-            torch.sum((v_i_quadruples[:, 2, :] ** 2), dim=1).reshape(-1, 1)
-            + torch.sum((v_i_quadruples[:, 3, :] ** 2), dim=1).reshape(1, -1)
-            - 2.0 * torch.mm(v_i_quadruples[:, 2, :], torch.transpose(v_i_quadruples[:, 3, :], 0, 1))
-        )
-        y_y_prime[y_y_prime < 0] = 0
-        x_y_prime = (
-            torch.sum((v_i_quadruples[:, 0, :] ** 2), dim=1).reshape(-1, 1)
-            + torch.sum((v_i_quadruples[:, 3, :] ** 2), dim=1).reshape(1, -1)
-            - 2.0 * torch.mm(v_i_quadruples[:, 0, :], torch.transpose(v_i_quadruples[:, 3, :], 0, 1))
-        )
-        x_y_prime[x_y_prime < 0] = 0
-        x_prime_y = (
-            torch.sum((v_i_quadruples[:, 1, :] ** 2), dim=1).reshape(-1, 1)
-            + torch.sum((v_i_quadruples[:, 2, :] ** 2), dim=1).reshape(1, -1)
-            - 2.0 * torch.mm(v_i_quadruples[:, 1, :], torch.transpose(v_i_quadruples[:, 2, :], 0, 1))
-        )
-        x_prime_y[x_prime_y < 0] = 0
-        # each inner product is a tensor of dimension len(v_i_quadruples) x len(v_i_quadruples), we return a
-        #  tensor of shape 4 x len(v_i_quadruples) x len(v_i_quadruples)
-        return torch.cat(
-            [x_x_prime.unsqueeze(0), y_y_prime.unsqueeze(0), x_y_prime.unsqueeze(0), x_prime_y.unsqueeze(0)], dim=0
-        )
-
-    def compute_euclidean_inner_products_all(self, X: torch.Tensor, Y: torch.Tensor) -> torch.Tensor:
-        # We want to compute the RBF kernel values without using the quadruples. To do this, we need to compute
-        # ||x - y||^2 for x and y. That is the inner product. Note that ||x - y||^2 = <x - y, x - y> = (x-y)^T(x-y)
+    def compute_euclidean_inner_products(self, X: torch.Tensor, Y: torch.Tensor) -> torch.Tensor:
+        # In this function, we assume that X, Y: n_samples, n_features
+        # We want to compute estimates of the expectation for each RBF kernel WITHOUT using a linear approximation.
+        # So we need to compute ||x - y||^2 for all pairs (x_j, x_k), (x_j, y_k), (x_k, y_j), and (y_j, y_k) for all
+        # j, k in range(n_samples).
+        # NOTE: ||x - y||^2 = <x - y, x - y> = <x, x> + <y, y> - 2<x, y>
         x_x_prime = (
             torch.sum((X**2), dim=1).reshape(-1, 1)
             + torch.sum((X**2), dim=1).reshape(1, -1)
             - 2.0 * torch.mm(X, torch.transpose(X, 0, 1))
         )
-        x_x_prime[x_x_prime < 0] = 0
```

Comment:
This is something that they added for two sample tests. I am not apposing to delete it though.